# Flipkart Analysis Project 
For any particular product category, is there an association rule between some of the product specification values and the product rating?

Project made by Kaushik,Mukund,Rajdeep,Salini , Batch 3, Globsyn Machine Learning with Python Summer School 2018

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('flipkart_spec_values_category_2.csv')

In [3]:
#dataset = dataset['product_specifications_values']
df.head()

Unnamed: 0 product_category  \
0           0        Clothing    
1           1       Furniture    
2           2        Footwear    
3           3        Clothing    
4           4    Pet Supplies    

                               product_specs_refined  \
0  {"product_specification"=>[{"key"=>"Number of ...   
1  {"product_specification"=>[{"key"=>"Installati...   
2  {"product_specification"=>[{"key"=>"Ideal For"...   
3  {"product_specification"=>[{"key"=>"Number of ...   
4  {"product_specification"=>[{"key"=>"Pet Type",...   

                       Product_specifications_values  
0  ['Pack of 3', 'Cotton Lycra', 'Cycling Shorts'...  
1  ['Installation and demo for this product is do...  
2  ['Women', 'Casual', 'Red', 'Patent Leather', '...  
3  ['Pack of 2', 'Cotton Lycra', 'Cycling Shorts'...  
4  ['Dog', 'Sicons', '500 ml', 'SH.DF-14', 'All P...

In [4]:
ind = df.columns

In [5]:
ind = list(ind)

In [6]:
ind[0] = 'notreq'

In [7]:
df.columns = ind

In [8]:
df.drop(['notreq'],axis=1,inplace=True)

In [9]:
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19441 entries, 0 to 19440
Data columns (total 3 columns):
product_category                 19441 non-null object
product_specs_refined            19441 non-null object
Product_specifications_values    19441 non-null object
dtypes: object(3)
memory usage: 227.9+ KB


In [10]:
type(df['product_category'][0])

str

In [11]:
df['product_category'][0]

'Clothing '

In [12]:
def trimstrings(x):
    return x.strip()

In [13]:
df['product_category'] = df['product_category'].apply(trimstrings)

In [14]:
df_clothing = df[df['product_category'] == 'Clothing']

In [15]:
df_clothing.head()

product_category                              product_specs_refined  \
0          Clothing  {"product_specification"=>[{"key"=>"Number of ...   
3          Clothing  {"product_specification"=>[{"key"=>"Number of ...   
6          Clothing  {"product_specification"=>[{"key"=>"Number of ...   
9          Clothing  {"product_specification"=>[{"key"=>"Number of ...   
11         Clothing  {"product_specification"=>[{"key"=>"Neck", "va...   

                        Product_specifications_values  
0   ['Pack of 3', 'Cotton Lycra', 'Cycling Shorts'...  
3   ['Pack of 2', 'Cotton Lycra', 'Cycling Shorts'...  
6   ['Pack of 4', 'Cotton Lycra', 'Cycling Shorts'...  
9   ['Pack of 4', 'Cotton Lycra', 'Cycling Shorts'...  
11  ['Round Neck', 'SwimLycra', 'Swim-dress', 'Pri...

In [16]:
#clothing_double_array = list(df_clothing['Product_specifications_values'])

In [17]:
#type(df_clothing['Product_specifications_values'][0])

In [18]:
#def convertlist(x):
    #return list(x)

In [19]:
#df_clothing['Product_specifications_values'] = df_clothing['Product_specifications_values'].apply(convertlist)

In [20]:
#clothing_double_array = list(df_clothing['Product_specifications_values'])

In [21]:
#clothing_double_array

In [22]:
'''
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(clothing_double_array).transform(clothing_double_array)
df = pd.DataFrame(te_ary, columns=te.columns_)
df
'''


'\nimport pandas as pd\nfrom mlxtend.preprocessing import TransactionEncoder\n\nte = TransactionEncoder()\nte_ary = te.fit(clothing_double_array).transform(clothing_double_array)\ndf = pd.DataFrame(te_ary, columns=te.columns_)\ndf\n'

In [23]:
'''from mlxtend.frequent_patterns import apriori

apriori(df, min_support=0)
'''

'from mlxtend.frequent_patterns import apriori\n\napriori(df, min_support=0)\n'

In [24]:
value_total_list = []

In [25]:
def getvalues(x):
    values = []
    value_count = x.count('value')
    for i in range(value_count):
        m = x.index('value')
        x = x[m+9:]
        m = x.index('"')
        y = x[:m]
        x = x[m:]
        values.append(y)
    return values

In [26]:
for row_index,row in df_clothing.iterrows():
    value_list = []
    try:
        value_list = getvalues(row['product_specs_refined'])
    except ValueError:
        value_list = 'No Values'
    value_total_list.append(value_list)

In [27]:
value_total_list

[['Pack of 3',
  'Cotton Lycra',
  'Cycling Shorts',
  'Solid',
  "Women's",
  'Gentle Machine Wash in Lukewarm Water, Do Not Bleach',
  'ALTHT_3P_21',
  '3 shorts'],
 ['Pack of 2',
  'Cotton Lycra',
  'Cycling Shorts',
  'Solid',
  "Women's",
  'Gentle Machine Wash in Lukewarm Water, Do Not Bleach',
  'ALTGHT_11',
  '2 shorts'],
 ['Pack of 4',
  'Cotton Lycra',
  'Cycling Shorts',
  'Solid',
  "Women's",
  '4 shorts',
  'ALTGHT4P_26',
  'Gentle Machine Wash in Lukewarm Water, Do Not Bleach'],
 ['Pack of 4',
  'Cotton Lycra',
  'Cycling Shorts',
  'Solid',
  "Women's",
  'Gentle Machine Wash in Lukewarm Water, Do Not Bleach',
  'ALTGHT4P_39',
  '4 shorts'],
 ['Round Neck',
  'SwimLycra',
  'Swim-dress',
  'Printed',
  "Women's",
  'Sports',
  'Wash with Similar Colors, Use Detergent for Colors',
  '1 Swimware'],
 ['Pack of 3',
  'Cotton Lycra',
  'Cycling Shorts',
  'Solid',
  "Women's",
  'Gentle Machine Wash in Lukewarm Water, Do Not Bleach',
  'ALTHT_3P_17',
  '3 shorts'],
 ['Pack o

In [28]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(value_total_list).transform(value_total_list)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

# 61-0288-001  #ZWSHRT006  \
0     False          False       False   
1     False          False       False   
2     False          False       False   
3     False          False       False   
4     False          False       False   
5     False          False       False   
6     False          False       False   
7     False          False       False   
8     False          False       False   
9     False          False       False   
10    False          False       False   
11    False          False       False   
12    False          False       False   
13    False          False       False   
14    False          False       False   
15    False          False       False   
16    False          False       False   
17    False          False       False   
18    False          False       False   
19    False          False       False   
20    False          False       False   
21    False          False       False   
22    False          False       False   
23    False          False       False   
24    False          False       False   
25    False          False       False   
26    False          False       False   
27    False          False       False   
28    False          False       False   
29    False          False       False   
...     ...            ...         ...   
5810  False          False       False   
5811  False          False       False   
5812  False          False       False   
5813  False          False       False   
5814  False          False       False   
5815  False          False       False   
5816  False          False       False   
5817  False          False       False   
5818  False          False       False   
5819  False          False       False   
5820  False          False       False   
5821  False          False       False   
5822  False          False       False   
5823  False          False       False   
5824  False          False       False   
5825  False          False       False   
5826  False          False       False   
5827  False          False       False   
5828  False          False       False   
5829  False          False       False   
5830  False          False       False   
5831  False          False       False   
5832  False          False       False   
5833  False          False       False   
5834  False          False       False   
5835  False          False       False   
5836  False          False       False   
5837  False          False       False   
5838  False          False       False   
5839  False          False       False   

      (5 Pockets) 2 Cross Pockets in front, 2 Bone Pockets at back and 1 Watch Pocket in front  \
0                                                 False                                          
1                                                 False                                          
2                                                 False                                          
3                                                 False                                          
4                                                 False                                          
5                                                 False                                          
6                                                 False                                          
7                                                 False                                          
8                                                 False                                          
9                                                 False                                          
10                                                False                                          
11                                                False                                          
12                                                False                                          
13                                  

In [29]:
from mlxtend.frequent_patterns import apriori

apriori(df, min_support=0.6)

support itemsets
0  0.628082   [1993]
1  0.676027   [5129]
2  0.609247   [6936]

In [30]:
apriori(df, min_support=0.6, use_colnames=True)

support     itemsets
0  0.628082     [Casual]
1  0.676027  [Pack of 1]
2  0.609247    [Women's]

In [31]:
frequent_itemsets = apriori(df, min_support=0.6, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support     itemsets  length
0  0.628082     [Casual]       1
1  0.676027  [Pack of 1]       1
2  0.609247    [Women's]       1

In [4]:
#oht_ary = te.fit(value_total_list).transform(value_total_list, sparse=True)
#sparse_df = pd.SparseDataFrame(te_ary, columns=te.columns_, default_fill_value=False)
#sparse_df

In [3]:
#apriori(sparse_df, min_support=0.6, use_colnames=True)

In [ ]:
frequent_itemsets[ (frequent_itemsets['length'] == 1) &
                   (frequent_itemsets['support'] >= 0.6) ]

### Apriori Algorithm predicted that Pack of 1 in clothing is the most favourable choice of the companies for selling.

In [34]:
frequent_itemsets['support'].max()

0.6760273972602739

In [35]:
frequent_itemsets[frequent_itemsets['support'] == frequent_itemsets['support'].max()]['itemsets']

1    [Pack of 1]
Name: itemsets, dtype: object

In [42]:
df_cleared = pd.read_csv('flipkart_final_cleared.csv')
df_cleared.head()

Unnamed: 0                           product_name  \
0           0    Alisha Solid Women's Cycling Shorts   
1           1    FabHomeDecor Fabric Double Sofa Bed   
2           2                             AW Bellies   
3           3    Alisha Solid Women's Cycling Shorts   
4           4  Sicons All Purpose Arnica Dog Shampoo   

                               product_category_tree  retail_price  \
0  ["Clothing >> Women's Clothing >> Lingerie, Sl...         999.0   
1  ["Furniture >> Living Room Furniture >> Sofa B...       32157.0   
2  ["Footwear >> Women's Footwear >> Ballerinas >...         999.0   
3  ["Clothing >> Women's Clothing >> Lingerie, Sl...         699.0   
4  ["Pet Supplies >> Grooming >> Skin & Coat Care...         220.0   

   discounted_price  is_FK_Advantage_product  product_rating  overall_rating  \
0             379.0                    False             5.0             3.0   
1           22646.0                    False             2.0             5.0   
2             499.0                    False             4.0             3.0   
3             267.0                    False             2.0             4.0   
4             210.0                    False             1.0             3.0   

    Brand_names product_category  \
0        Alisha        Clothing    
1  FabHomeDecor       Furniture    
2            AW        Footwear    
3        Alisha        Clothing    
4        Sicons    Pet Supplies    

                               product_specs_refined  
0  {"product_specification"=>[{"key"=>"Number of ...  
1  {"product_specification"=>[{"key"=>"Installati...  
2  {"product_specification"=>[{"key"=>"Ideal For"...  
3  {"product_specification"=>[{"key"=>"Number of ...  
4  {"product_specification"=>[{"key"=>"Pet Type",...

In [46]:
df_clothing_with_rating = df_cleared[df_cleared['product_category']=='Clothing ']

In [66]:
value_list_with_rating = []
for row_index,row in df_clothing_with_rating.iterrows():
    value_list = []
    try:
        value_list = getvalues(row['product_specs_refined'])
    except ValueError:
        value_list = ['No Values']
    value_list.append(str(int(row['product_rating'])))
    value_list_with_rating.append(value_list)

In [67]:
value_list_with_rating[-4]

['Solid',
 'Casual',
 "Women's",
 '3/4 Sleeve',
 'Pack of 1',
 'Slim',
 'Georgette',
 'Slim',
 'STARSAT-PICH',
 '5']

In [68]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(value_list_with_rating).transform(value_list_with_rating)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

# 61-0288-001  #ZWSHRT006  \
0             False       False   
1             False       False   
2             False       False   
3             False       False   
4             False       False   
5             False       False   
6             False       False   
7             False       False   
8             False       False   
9             False       False   
10            False       False   
11            False       False   
12            False       False   
13            False       False   
14            False       False   
15            False       False   
16            False       False   
17            False       False   
18            False       False   
19            False       False   
20            False       False   
21            False       False   
22            False       False   
23            False       False   
24            False       False   
25            False       False   
26            False       False   
27            False       False   
28            False       False   
29            False       False   
...             ...         ...   
5810          False       False   
5811          False       False   
5812          False       False   
5813          False       False   
5814          False       False   
5815          False       False   
5816          False       False   
5817          False       False   
5818          False       False   
5819          False       False   
5820          False       False   
5821          False       False   
5822          False       False   
5823          False       False   
5824          False       False   
5825          False       False   
5826          False       False   
5827          False       False   
5828          False       False   
5829          False       False   
5830          False       False   
5831          False       False   
5832          False       False   
5833          False       False   
5834          False       False   
5835          False       False   
5836          False       False   
5837          False       False   
5838          False       False   
5839          False       False   

      (5 Pockets) 2 Cross Pockets in front, 2 Bone Pockets at back and 1 Watch Pocket in front  \
0                                                 False                                          
1                                                 False                                          
2                                                 False                                          
3                                                 False                                          
4                                                 False                                          
5                                                 False                                          
6                                                 False                                          
7                                                 False                                          
8                                                 False                                          
9                                                 False                                          
10                                                False                                          
11                                                False                                          
12                                                False                                          
13                                                False                                          
14                                                False                                          
15                                                False                                          
16                                                False                                          
17                                                False                

In [71]:
from mlxtend.frequent_patterns import apriori

apriori(df, min_support=0.4)

support      itemsets
0  0.628082        [1993]
1  0.676027        [5129]
2  0.425342        [6146]
3  0.609247        [6935]
4  0.463527  [1993, 5129]
5  0.430479  [5129, 6935]

In [75]:
apriori(df, min_support=0.4, use_colnames=True)

support              itemsets
0  0.628082              [Casual]
1  0.676027           [Pack of 1]
2  0.425342               [Solid]
3  0.609247             [Women's]
4  0.463527   [Casual, Pack of 1]
5  0.430479  [Pack of 1, Women's]

In [77]:
apriori(df, min_support=0.4, use_colnames=True)

support              itemsets
0  0.628082              [Casual]
1  0.676027           [Pack of 1]
2  0.425342               [Solid]
3  0.609247             [Women's]
4  0.463527   [Casual, Pack of 1]
5  0.430479  [Pack of 1, Women's]

In [81]:
frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                               itemsets  length
0    0.148288                                    [0]       1
1    0.174315                                    [1]       1
2    0.155822                                    [2]       1
3    0.177397                                    [3]       1
4    0.176712                                    [4]       1
5    0.181336                                    [5]       1
6    0.628082                               [Casual]       1
7    0.392466                               [Cotton]       1
8    0.172432                          [Full Sleeve]       1
9    0.175514                          [Half Sleeve]       1
10   0.279623                                [Men's]       1
11   0.166610                                   [No]       1
12   0.676027                            [Pack of 1]       1
13   0.220548                              [Printed]       1
14   0.364041                              [Regular]       1
15   0.211815                           [Round Neck]       1
16   0.128596                           [Sleeveless]       1
17   0.425342                                [Solid]       1
18   0.137671                             [Wirefree]       1
19   0.609247                              [Women's]       1
20   0.100171                         [0, Pack of 1]       2
21   0.106678                            [1, Casual]       2
22   0.114212                         [1, Pack of 1]       2
23   0.104966                           [1, Women's]       2
24   0.105993                         [2, Pack of 1]       2
25   0.113185                            [3, Casual]       2
26   0.116438                         [3, Pack of 1]       2
27   0.110616                           [3, Women's]       2
28   0.108219                            [4, Casual]       2
29   0.118836                         [4, Pack of 1]       2
..        ...                                    ...     ...
72   0.112671                [Casual, Cotton, Men's]       3
73   0.203425            [Casual, Cotton, Pack of 1]       3
74   0.144349              [Casual, Cotton, Regular]       3
75   0.115582                [Casual, Cotton, Solid]       3
76   0.152740              [Casual, Cotton, Women's]       3
77   0.115753           [Casual, Half Sleeve, Men's]       3
78   0.126027       [Casual, Half Sleeve, Pack of 1]       3
79   0.153082             [Casual, Men's, Pack of 1]       3
80   0.130822           [Casual, Pack of 1, Printed]       3
81   0.207363           [Casual, Pack of 1, Regular]       3
82   0.144178        [Casual, Pack of 1, Round Neck]       3
83   0.187158             [Casual, Pack of 1, Solid]       3
84   0.292637           [Casual, Pack of 1, Women's]       3
85   0.121233               [Casual, Regular, Solid]       3
86   0.174315             [Casual, Regular, Women's]       3
87   0.178767               [Casual, Solid, Women's]       3
88   0.115753             [Cotton, Men's, Pack of 1]       3
89   0.133904           [Cotton, Pack of 1, Regular]       3
90   0.146918           [Cotton, Pack of 1, Women's]       3
91   0.102226             [Cotton, Regular, Women's]       3
92   0.110103        [Half Sleeve, Men's, Pack of 1]       3
93   0.117123            [Pack of 1, Regular, Solid]       3
94   0.175171          [Pack of 1, Regular, Women's]       3
95   0.189897            [Pack of 1, Solid, Women's]       3
96   0.113356              [Regular, Solid, Women's]       3
97   0.114041           [Regular, Wirefree, Women's]       3
98   0.102055   [Casual, Cotton, Pack of 1, Regular]       4
99   0.104452   [Casual, Cotton, Pack of 1, Women's]       4
100  0.126712  [Casual, Pack of 1, Regular, Women's]       4
101  0.125171    [Casual, Pack of 1, Solid, Women's]       4

[102 rows x 3 columns]

In [86]:
frequent_itemsets[ (frequent_itemsets['length'] == 2) &
                   (frequent_itemsets['support'] >= 0.4) ]

support              itemsets  length
39  0.463527   [Casual, Pack of 1]       2
63  0.430479  [Pack of 1, Women's]       2

# This shows that the casual pack of 1 selling is the most frequent choice of flipkart

In [112]:
itemsets = frequent_itemsets[ (frequent_itemsets['length'] == 2) &
                   (frequent_itemsets['support'] >= 0.1) ]

In [117]:
type(itemsets['itemsets'])

pandas.core.series.Series

In [118]:
def to_string(x):
    return str(x)

In [119]:
itemsets['itemsets'].apply(to_string)

20              ['0', 'Pack of 1']
21                 ['1', 'Casual']
22              ['1', 'Pack of 1']
23                ['1', "Women's"]
24              ['2', 'Pack of 1']
25                 ['3', 'Casual']
26              ['3', 'Pack of 1']
27                ['3', "Women's"]
28                 ['4', 'Casual']
29              ['4', 'Pack of 1']
30                ['4', "Women's"]
31                 ['5', 'Casual']
32              ['5', 'Pack of 1']
33                ['5', "Women's"]
34            ['Casual', 'Cotton']
35       ['Casual', 'Full Sleeve']
36       ['Casual', 'Half Sleeve']
37             ['Casual', "Men's"]
38                ['Casual', 'No']
39         ['Casual', 'Pack of 1']
40           ['Casual', 'Printed']
41           ['Casual', 'Regular']
42        ['Casual', 'Round Neck']
43             ['Casual', 'Solid']
44           ['Casual', "Women's"]
45             ['Cotton', "Men's"]
46         ['Cotton', 'Pack of 1']
47           ['Cotton', 'Printed']
48           ['Cotto

In [120]:
itemsets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43 entries, 20 to 71
Data columns (total 3 columns):
support     43 non-null float64
itemsets    43 non-null object
length      43 non-null int64
dtypes: float64(1), int64(1), object(1)
memory usage: 2.2+ KB


In [151]:
rule_list_total = []
for row_index,row in itemsets.iterrows():
    rule_list = []
    if row['itemsets'][0] in ['0','1','2','3','4','5']:
        rule_list.append(str(row['itemsets']))
        rule_list.append(row['support'])
        rule_list_total.append(rule_list)

In [152]:
rule_list_total

[["['0', 'Pack of 1']", 0.10017123287671233],
 ["['1', 'Casual']", 0.10667808219178082],
 ["['1', 'Pack of 1']", 0.11421232876712328],
 ['[\'1\', "Women\'s"]', 0.10496575342465754],
 ["['2', 'Pack of 1']", 0.10599315068493151],
 ["['3', 'Casual']", 0.11318493150684932],
 ["['3', 'Pack of 1']", 0.11643835616438356],
 ['[\'3\', "Women\'s"]', 0.11061643835616439],
 ["['4', 'Casual']", 0.10821917808219178],
 ["['4', 'Pack of 1']", 0.11883561643835616],
 ['[\'4\', "Women\'s"]', 0.10308219178082192],
 ["['5', 'Casual']", 0.12037671232876712],
 ["['5', 'Pack of 1']", 0.12465753424657534],
 ['[\'5\', "Women\'s"]', 0.11044520547945205]]

In [157]:
rule_df_clothing = pd.DataFrame(data=rule_list_total,columns=['rating_specificationvalue','support'])

In [158]:
rule_df_clothing.head()

rating_specificationvalue   support
0        ['0', 'Pack of 1']  0.100171
1           ['1', 'Casual']  0.106678
2        ['1', 'Pack of 1']  0.114212
3          ['1', "Women's"]  0.104966
4        ['2', 'Pack of 1']  0.105993

In [160]:
rule_df_clothing.max()

rating_specificationvalue    ['5', 'Pack of 1']
support                                0.124658
dtype: object

## Pack of 1 with a rating of 5 gets a support value of 0.124658